City-of-Helsinki/hel-fi-drupal-grants
# Feature/AU-1408: Webform configuration override

# [AU-1408](https://helsinkisolutionoffice.atlassian.net/browse/AU-1408)

## What was done
This pull request implements a new drush command in the `WebformConfigOverrideCommands.php` file. This command makes it possible to import Webform configuration overrides. The overrides should be defined in the `grants_metadata.settings.yml` file. The command can be executed by running `drush gwco`.

## How to install

Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-1408-webform-config-override`
  * `make fresh`
  * `make drush-cr`

## How to test

Start by defining the overrides you want to test in the `grants_metadata.settings.yml` file. Any implemented overrides need to be defined under a `overridden_configuration` key. Each override should be indexed with the forms **application type ID**. The overrides should be structured in the same manner as the original Webform configurations and can therefore be defined in the following manner:

**Example (all possible overrides)**

```
overridden_configuration:
  - 53:
      grants_metadata:
        applicationTypeSelect: '53'
        applicationType: KASKOIPLISA
        applicationTypeID: '53'
        applicationIndustry: KASKO
        applicantTypes:
          registered_community: registered_community
        applicationTypeTerms:
          56: '56'
        applicationTargetGroup: '21'
        applicationOpen: '2023-06-16T08:30:52'
        applicationClose: '2023-08-19T08:36:07'
        applicationActingYears: {  }
        applicationContinuous: 0
        disableCopying: 0
        applicationActingYearsType: current_and_next_x_years
        applicationActingYearsNextCount: '1'
  - 47:
      grants_metadata:
        applicationType: KUVATOIM
        applicationTypeID: '47'
        applicantTypes:
          registered_community: registered_community
        applicationTypeTerms:
          45: '45'
        applicationOpen: '2022-10-02T11:08:26'
        applicationClose: '2023-07-31T11:19:00'
        applicationContinuous: 0
        applicationTargetGroup: '22'
        disableCopying: 0
        applicationTypeSelect: '47'
        applicationIndustry: KUVA
        applicationActingYears:
          2023: '2023'
          2024: '2024'
          2025: '2025'
```

The provided overrides are not verified in any way (the implementation uses a simple `array_merge`), so one should be careful when implementing these changes. A more realistic set of overrides would therefore look something like this:

```
overridden_configuration:
  - 53:
      grants_metadata:
        applicationOpen: '2022-10-02T11:08:26'
        applicationClose: '2025-07-31T11:19:00'
        disableCopying: 1
  - 47:
      grants_metadata:
        applicationOpen: '2022-10-02T11:08:26'
        applicationClose: '2026-07-31T11:19:00'
        disableCopying: 1
```

After this, simply run `drush gwco` and verify that the overrides worked. If you which to view a debugging output, then simply uncomment the lines **103-105** in `WebformConfigOverrideCommands.php`. If you manage to break something, then a simple `drush gwi` should revert any changes.

## Things to note
The new Webform "versioning" that we are implementing might require code changes to this implementation. Further investigation needed when said features are in place.

[AU-1408]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1408?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
